foxai/explainer/fullgradeints.py

```diff
@@ -0,0 +1,462 @@
+
+        Returns:
+            input image gradients (contribution of each input pixel to the final output)
+            list of contributions of bias terms in each layer in the model (each layer with the bias term)
+        """
+
+        self.model.eval()
+        image = image.requires_grad_()
+        out = self.model(image)
+
+        if target_class is None:
+            target_class = out.data.max(1, keepdim=True)[1]
+
+        # Select the output unit corresponding to the target class
+        # -1 compensates for negation in nll_loss function
+        output_scalar = -1.0 * F.nll_loss(out, target_class.flatten(), reduction="sum")
+
+        input_gradient, feature_gradients = self.model_ext.get_feature_grads(
+            image, output_scalar
+        )
+
+        # Compute feature-gradients \times bias
+        bias_times_gradients: List[torch.Tensor] = []
+        length = len(self._biases)
+
+        for i in range(length):
+
+            # feature gradients are indexed backwards
+            # because of backprop
+            g = feature_gradients[length - 1 - i]
+
+            # reshape bias dimensionality to match gradients
+            bias_size = [1] * len(g.size())
+            bias_size[1] = self._biases[i].size(0)
+            b = self._biases[i].view(tuple(bias_size))
+
+            bias_times_gradients.append(g * b.expand_as(g))
+
+        return input_gradient, bias_times_gradients
+
+    def _post_process(
+        self, input_tensor: torch.Tensor, eps: float = 1e-6
+    ) -> torch.Tensor:
+        """Normalize gradient maps.
+
+        Args:
+            input_tensor: the gradient map to be normalized
+            eps: epsilon
+
+        Return:
+            normalized map
+        """
+
+        # Absolute value
+        input_tensor = abs(input_tensor)
+
+        # Rescale operations to ensure gradients lie between 0 and 1
+        flatin = input_tensor.view((input_tensor.size(0), -1))
+        temp, _ = flatin.min(1, keepdim=True)
+        input_tensor = input_tensor - temp.unsqueeze(1).unsqueeze(1)
+
+        flatin = input_tensor.view((input_tensor.size(0), -1))
+        temp, _ = flatin.max(1, keepdim=True)
+        input_tensor = input_tensor / (temp.unsqueeze(1).unsqueeze(1) + eps)
+        return input_tensor
+
+    def compute_saliency(
+        self,
+        input_data: torch.Tensor,
+        target: Optional[Union[torch.Tensor, int]] = None,
+    ) -> torch.Tensor:
+        """Compute FullGrad saliency map.
+
+        The function combines input gradient map with bias layer gradient maps.
+
+        Args:
+            input_data: the input image to the model
+            target: the target class to be explained. If none the argmax class is chosen.
+
+        Returns:
+            silency map
+        """
+        if isinstance(target, int):
+            target = torch.tensor(data=[target] * input_data.shape[0]).to(self._device)
+
+        self.model.eval()
```

Comment:
Same as above, we should somehow revert this change to the model.

Reply:
done